src/nsmgr/internal/pkg/deviceplugin/server.go

```diff
@@ -0,0 +1,251 @@
+	allocatedDevices      map[string]*pluginapi.Device
+	executor              serialize.Executor
+	listAndWatchListeners []pluginapi.DevicePlugin_ListAndWatchServer
+	grpcServer            *grpc.Server
+	listenEndpoint        string
+	sock                  net.Listener
+	values                *flags.DefinedFlags
+}
+
+func (n *nsmgrDevicePlugin) ListenEndpoint() string {
+	return n.listenEndpoint
+}
+func (n *nsmgrDevicePlugin) Stop() {
+	if n.sock != nil {
+		_ = n.sock.Close()
+		n.sock = nil
+	}
+	if n.grpcServer != nil {
+		n.grpcServer.Stop()
+		n.grpcServer = nil
+	}
+}
+func (n *nsmgrDevicePlugin) Start() error {
+	// We do not need a peer tracker in case of single source GRPC server.
+	_ = os.Remove(n.listenEndpoint)
+	var err error
+	n.sock, err = net.Listen("unix", n.listenEndpoint)
+	if err != nil {
+		err = errors.WithMessagef(err, "failed to listen on %s: %+v", n.listenEndpoint, err)
+		return err
+	}
+
+	n.grpcServer = grpc.NewServer()
+	pluginapi.RegisterDevicePluginServer(n.grpcServer, n)
+	go func() {
+		_ = n.grpcServer.Serve(n.sock)
+	}()
+
+	return n.Register()
+}
+
+func NewServer(values *flags.DefinedFlags) NsmDevicePluginServer {
+	rv := &nsmgrDevicePlugin{
+		devices:          make(map[string]*pluginapi.Device, constants.DeviceBuffer),
+		allocatedDevices: make(map[string]*pluginapi.Device, constants.DeviceBuffer),
+		executor:         serialize.NewExecutor(),
+		listenEndpoint:   values.DeviceAPIListenEndpoint,
+		values:           values,
+	}
+	rv.resizeDevicePool()
+	return rv
+}
+
+func (n *nsmgrDevicePlugin) ListAndWatch(e *pluginapi.Empty, s pluginapi.DevicePlugin_ListAndWatchServer) error {
+	n.executor.AsyncExec(func() {
+		n.listAndWatchListeners = append(n.listAndWatchListeners, s)
+		listAndWatchResponse := &pluginapi.ListAndWatchResponse{}
+		for _, device := range n.devices {
+			listAndWatchResponse.Devices = append(listAndWatchResponse.Devices, device)
+		}
+		_ = s.Send(listAndWatchResponse)
+	})
+
+	<-s.Context().Done()
+	n.executor.AsyncExec(func() {
+		var listAndWatchListeners []pluginapi.DevicePlugin_ListAndWatchServer
+		for _, listAndWatchListener := range n.listAndWatchListeners {
+			if listAndWatchListener != s {
+				listAndWatchListeners = append(listAndWatchListeners, listAndWatchListener)
+			}
+		}
+		// Update list
+		n.listAndWatchListeners = listAndWatchListeners
+	})
+	return nil
+}
+
+func (n *nsmgrDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+	containerResponses := []*pluginapi.ContainerAllocateResponse{}
+	for _, req := range reqs.GetContainerRequests() {
+		for _, deviceid := range req.GetDevicesIDs() {
+			_, ok := n.devices[deviceid]
+			if !ok {
+				return nil, fmt.Errorf("device id passed not found %v", deviceid)
+			}
+			// Clean any memif files, or endpoint socket files.
+			// Connections will be closed automatically.
+			n.cleanFolder(ctx, hostDeviceDirectory(n.values, deviceid))
```

Comment:
hostDeviceDirectory is only used in Allocate, would it not make more sense to replace it with:
```
hostDeviceDirectory := path.Join(values.BaseDir,deviceId)
```

and then subsequently use that variable?